Update for Prism V3

wd-mgreynolds opened the pull request.

CurtLH commented:
@wd-mgreynolds -- I tried to install your fork of the library, but the following error was generated:
```
Traceback (most recent call last):
  File "/Users/curtis.hampton/miniconda3/envs/example-dev/bin/prism", line 5, in <module>
    from prism.cli import main
  File "/Users/curtis.hampton/miniconda3/envs/example-dev/lib/python3.8/site-packages/prism/cli.py", line 9, in <module>
    import commands.tables_commands as t_commands
ModuleNotFoundError: No module named 'commands'
```

I wonder if you might have to edit the `__init__.py` to include the modules within `./commands/`?

CurtLH commented on requirements.txt:
```diff
@@ -1,2 +1,4 @@
 click>=6
 requests>=2.20.0,<3
+configparser>=6
```
I believe `configparser` is a built-in Python module.  Why does it need to be explicitly included here for install?

wd-mgreynolds replied:
Removed.

CurtLH commented on requirements.txt:
```diff
@@ -1,2 +1,4 @@
 click>=6
 requests>=2.20.0,<3
+configparser>=6
+pandas~=2.0.0
```
It looks like pandas is only used for the `json_normalize()` function.  Pandas is rather large dependency, that doesn't always play well other dependencies, so I'd prefer to not include it unless there is some compelling reason that can't be overcome with the existing dependencies. Maybe we could write our own function to flatten JSON data?

wd-mgreynolds replied:
Agreed - will remove on next push.

wd-mgreynolds replied:
I decided to remove the tabular output feature throughout.  Nobody asked for it (except me) and it will not be missed.  Also, dumping Prism JSON to CSV would be a pain.

CurtLH commented on prism/commands/wql_commands.py:
I'm not a big fan of including WQL in this package.  Given the package is called `prism-python`, I would think everything would relate to Prism, and while I understand there is some overlap with WQL, it seems to me that should be a separate package.  What do you think?

wd-mgreynolds replied:
Understand the concern and will remove - there is always the official "wcpcli wql:query" to do WQL.  However, I do like my addition to automatically expand "select *" to all the columns (i'll keep a backup of that code).

CurtLH commented on prism/commands/raas_commands.py:
I'm not a big fan of including RaaS in this package, especially as there the [raas-python](https://github.com/Workday/raas-python) package already.  Given the package is called `prism-python`, I would think everything would relate to Prism, and while I understand there is some overlap with RaaS, it seems to me that should be a separate package.  What do you think?

wd-mgreynolds replied:
Understand the concern and will remove.  I included it here to cut down on the jumble of tools we provide, but keeping it in this tool does dilute the focus.

CurtLH reviewed (commented):
@wd-mgreynolds -- this isn't a complete review, but a couple things that we can start to think about as we review all the proposed changes.

CurtLH commented:
@wd-mgreynolds -- can you remove Python 3.7 from the versions to test the build against in the [GitHub workflow](https://github.com/wd-mgreynolds/prism-python/blob/master/.github/workflows/pythonpackage.yml#L11)?  

Python 3.7 reached end of life on 2023-06-27.

CurtLH commented on .github/workflows/pythonpackage.yml:
```diff
@@ -8,7 +8,7 @@ jobs:
     strategy:
       matrix:
-        python-version: [3.7, 3.8, 3.9]
+        python-version: [3.9]
```
Thanks for dropping 3.7, but I don't think we should drop 3.8 as well, as that will still be supported for another year at this point.

wd-mgreynolds replied:
Any thoughts on the "classifiers" list in setup.py?  The value in the original repository is:

   classifiers=["Programming Language :: Python :: 3.6", "Programming Language :: Python :: 3.7"]

CurtLH commented on prism/__init__.py:
```diff
@@ -1,8 +1,8 @@
-from prism.prism import Prism, load_schema, create_table, upload_file
```
I see the functions `load_schema()`, `create_table()`, and `upload_file()` were dropped in your suggested changes.  Those functions were essential in my mind as not all users want to go through every single step needed just to create a table or upload a file.  What are your thoughts behind removing these wrapper functions that simplify the process of interacting with the API?

wd-mgreynolds replied:
I was thinking about that today and was wondering about those routines in the prism class and when they would be used.  I was leaning towards wrapping those activities exclusively in the CLI.  However if we expect the prism class to be part of other projects, I can quickly refactor the code for loading files that I have in the tables, buckets and dataChange commands into callable functions similar to the original upload function.  This is a minor refactor where the CLI commands would simply call the functions in the prism class.  Note: each of my upload routines is doing different sanity checks so I'd likely end up with more than one function to handle uploads.  I'll review the original functions to ensure no features/capabilities are lost.

CurtLH commented on prism/prism.py:
```diff
@@ -98,364 +393,1074 @@ def create_bearer_token(self):
+
+        # Start setting up the API call parameters.
+        params = {
+            'limit': limit if limit is not None else 100,
```
For all of these parameters, why not just set the default values when defining the function?

wd-mgreynolds replied:
I'm using None as a cheap way to tell if the user wants everything, i.e., no limit / all items.  There is a hole in my logic I need to think about - standby.

wd-mgreynolds replied:
So, when the caller does NOT pass a values, i.e., None, this means get all.  I thought this would be better than using, say "-1" or "999" as a limit value - especially since these are invalid on the real API call (don't want to confuse the user).  When the user does a non-table search call and does provide a limit, this routine works exactly like standard API call.  **The value-add here is I don't force them to accept a limit.**

If the user does a table search, limits are ignored because I have to pass all the pages of tables to find matches.  Also, if I tried to apply limits to a search for table names then I'm just replicating the API limit/offset functionality.

**Question**: As I think about this, this functionality may be better placed in the cli.py + tables_commands.py and not in prism.py class, but it seem generically useful.  Any thoughts on moving the logic out of prism.py?

CurtLH replied:
I see.  I think it's fine to keep it here.

CurtLH commented on prism/prism.py:
```diff
@@ -3,79 +3,375 @@
+
+
+def buckets_gen_name():
+    bucket_name = "cli_" + uuid.uuid4().hex
```
Per advice for the Workday-on-Workday team, they would like the bucket name to start with `prism_python_` so they can audit how this library is being used.  `cli_` feels too generic and not clear that it relates to this interface of the Prism API.

wd-mgreynolds replied:
Done.

CurtLH commented on setup.py:
```diff
@@ -12,10 +12,10 @@
     packages=["prism"],
-    package_data={"prism": ["data/*"]},
-    entry_points={"console_scripts": ["prism=prism.cli:main"]},
+    package_data={"prism": ["data/*", "commands/*"]},
```
I don't believe we need to include `"commands/*"` here, as `package_data` is meant to tell setuptools to include the data files in this directory when building the package.  This is necessary to the data test files are included when the package builds through the CI platform.

wd-mgreynolds replied:
When I don't include the directory, I get errors on when I run: 

[ec2-user@ip-172-31-26-32 ~]$ prism tables get
Traceback (most recent call last):
  File "/home/ec2-user/.local/bin/prism", line 5, in <module>
    from prism.cli import cli
  File "/home/ec2-user/.local/lib/python3.9/site-packages/prism/cli.py", line 9, in <module>
    from .commands import tables_commands as t_commands
**ModuleNotFoundError: No module named 'prism.commands'**

CurtLH replied:
Check out this [documentation](https://packaging.python.org/en/latest/guides/distributing-packages-using-setuptools/#package-data).  I don't believe `pacakge_data` is the right way to solve that issue. 

> Often, additional files need to be installed into a [package](https://packaging.python.org/en/latest/glossary/#term-Import-Package). These files are often data that’s closely related to the package’s implementation, or text files containing documentation that might be of interest to programmers using the package. These files are called “package data”.

Instead, I think we need to explicitly tell the package where `prism.commands` is within the `__init__.py` file.  Check out this [documentation](https://docs.python.org/3.8/tutorial/modules.html#packages) for an example.

CurtLH commented:
@wd-mgreynolds -- the continuous integration testing is currently failing because there are a number of files that need to be reformatted using the style dictated within the package [black](https://black.readthedocs.io/en/stable/).  When you run the tool `black`, you will want to be in the root directory of the repo so that the [pyproject.toml](https://github.com/Workday/prism-python/blob/master/pyproject.toml) gets picked up and the specified configuration gets applied.

wd-mgreynolds commented:
On the latest push, the black, flake and pytest workflow steps succeed locally.

CurtLH commented:
@wd-mgreynolds -- I'm able to execute the command `prism --help` and it works as expected, but I get the following error with `prism tables --help`.

```prism tables --help
Traceback (most recent call last):
  File "/Users/curtis.hampton/miniconda3/envs/prism-dev/bin/prism", line 8, in <module>
    sys.exit(cli())
  File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/core.py", line 1256, in invoke
    Command.invoke(self, ctx)
  File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
TypeError: cli() missing 2 required positional arguments: 'username' and 'password'
```

I have `config.ini` in the same directory and it looks like this:
```prism_client_id=**********
prism_client_secret==**********
prism_refresh_token==**********
workday_base_url==**********
workday_tenant_name==**********
```

CurtLH commented:
@wd-mgreynolds -- we also need to ensure the README is updated with basic examples of how to create a table and upload a file, using both the Python package and the CLI. I don't think we need to go into more detail than that, as we can do that within the future documentation.  But those are the two most basic and common uses someone would use this package for, so that should be included in the README.

wd-mgreynolds commented:
Fixed - there was a left-over garbage after a code refactor - fixed.  This begs the question - should the pytest be more extensive?

I also added a sanity check to make sure a configuration (base_url/tenant/id/secret) was detected.

> @wd-mgreynolds -- I'm able to execute the command `prism --help` and it works as expected, but I get the following error with `prism tables --help`.
> 
> ```
> Traceback (most recent call last):
>   File "/Users/curtis.hampton/miniconda3/envs/prism-dev/bin/prism", line 8, in <module>
>     sys.exit(cli())
>   File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/core.py", line 829, in __call__
>     return self.main(*args, **kwargs)
>   File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/core.py", line 782, in main
>     rv = self.invoke(ctx)
>   File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/core.py", line 1256, in invoke
>     Command.invoke(self, ctx)
>   File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/core.py", line 610, in invoke
>     return callback(*args, **kwargs)
>   File "/Users/curtis.hampton/miniconda3/envs/prism-dev/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
>     return f(get_current_context(), *args, **kwargs)
> TypeError: cli() missing 2 required positional arguments: 'username' and 'password'
> ```
> 
> I have `config.ini` in the same directory and it looks like this:
> 
> ```
> prism_client_secret==**********
> prism_refresh_token==**********
> workday_base_url==**********
> workday_tenant_name==**********
> ```

CurtLH commented:
> Fixed - there was a left-over garbage after a code refactor - fixed. This begs the question - should the pytest be more extensive?
> 
> I also added a sanity check to make sure a configuration (base_url/tenant/id/secret) was detected.
> 

@wd-mgreynolds -- one day, I do think pytest could be more extensive.  It's a bit tricky through because we would need to create a mock API in order for it to interact with.  There are things like [requests-mock](https://requests-mock.readthedocs.io/en/latest/) and [vcrpy](https://vcrpy.readthedocs.io/en/latest/) to do this, but let's explore those after we get this initial support for V3 merged.

wd-mgreynolds commented:
I've pushed a first-cut of the revised README.md.  I elaborated quite a bit with the idea that a non-administrator would be looking for this type of functionality, so more info is better.  This is not final, but the overall structure is in place.  Please review for structure and level of detail - final word smithing is still on the plate.

wd-mgreynolds commented:
I did another pass on the README - the text should be complete.  I'm still thinking about the images, in terms of size, borders, and possibly animated GIFs.  I like the idea of animations, but it can get really annoying after the first time the animation runs.

CurtLH commented:
> I did another pass on the README - the text should be complete. I'm still thinking about the images, in terms of size, borders, and possibly animated GIFs. I like the idea of animations, but it can get really annoying after the first time the animation runs.

@wd-mgreynolds -- I appreciate the thoroughness, but I think this is too much detail for the README. This level of documentation would be fine in Community, but I don't think GitHub is the right place.  For example, check out the [README for the Google API Client](https://github.com/googleapis/google-api-python-client#google-api-client) or the [README for Boto3](https://github.com/boto/boto3#boto3---the-aws-sdk-for-python).  

I think this README should have the following sections:

- Overview of the package
- Description of why it's useful
- How to install it (this should only include the pip installation instructions, we don't need to show people how to install Python, PIP, or Git)
- Examples of how to use the package (both the CLI and the Python Package)
- How to report bugs or request new features

After this initial release, we can consider building further documentation helping a user get started, but the README should be brief and only the essential information.

wd-mgreynolds commented:
Thanks for the feedback and I appreciate your partnership.  Unfortunately other commitments will limit the speed with which I will be able to make further changes.  My manager has requested a hand-off meeting to provide any last details or answer any questions.  Let me know if there's a better time next week and I'll throw a meeting on the calendar.

CurtLH commented:
@wd-mgreynolds -- if you want to add me as a contributor to your forked repo, I can make some edits to the README.
